### PR TITLE
Hare bench

### PR DESCRIPTION
**Pull request by gavraz**

### CODE REVIEW

File: oracle/oracle_client.go [hunk 1]
```diff
@@ -67,7 +67,7 @@ func (hr *HTTPRequester) Get(api, data string) []byte {
 
 // OracleClient is a temporary replacement fot the real oracle. its gets accurate results from a server.
 type OracleClient struct {
-	world  uint64
+	world  int
```
Comment (antonlerner) [inline oracle/oracle_client.go]: why do you care?

Comment (gavraz) [reply oracle/oracle_client.go]: There is a bug in the Cobra package. On int it doesn't happen.

Comment (antonlerner) [reply oracle/oracle_client.go]: just cast to int when needed

File: app/cmd/hare.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	app.oracle = oracle.NewOracleClientWithWorldID(app.Config.OracleServerWorldId)
+	app.oracle.Register(true, pub.String()) // todo: configure no faulty nodes
+	hareOracle := oracle.NewHareOracleFromClient(app.oracle)
+
+	broker := hare.NewBroker(swarm, hare.NewEligibilityValidator(hare.NewHareOracle(hareOracle, app.Config.HARE.N), lg))
+	app.broker = broker
+	broker.Start()
+	app.p2p.Start()
+
+	time.Sleep(10 * time.Second)
```
Comment (antonlerner) [inline app/cmd/hare.go]: why do you need this sleep? you can export waitForBootstrap and use it

File: app/cmd/hare.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+	proc := hare.NewConsensusProcess(app.Config.HARE, 1, buildSet(), hareOracle, app.sgn, swarm, make(chan hare.TerminationOutput, 1), lg)
+	app.proc = proc
+	proc.SetInbox(broker.Register(proc.Id()))
+	proc.Start()
```
Comment (antonlerner) [inline app/cmd/hare.go]: is this blocking? should it be?

File: app/cmd/hare.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+func (app *HareApp) Cleanup() {
+	// TODO: move to array of cleanup functions and execute all here
+	app.oracle.Unregister(true, app.sgn.Verifier().String())
+}
+
+func buildSet() *hare.Set {
+	s := hare.NewEmptySet(200)
+
+	for i := 0; i < 200; i++ {
```
Comment (antonlerner) [inline app/cmd/hare.go]: what is this 200? use const or config

Comment (gavraz) [reply app/cmd/hare.go]: Fixed

File: oracle/oracle_client.go (hunk 1 above)
Comment (gavraz) [reply oracle/oracle_client.go]: Fixed

File: app/cmd/app_test.go [hunk 5]
```diff
@@ -54,9 +54,12 @@ func TestApp(t *testing.T) {
 	// remove all injected test flags for now
 	os.Args = []string{"/go-spacemesh", "--json-server=true"}
 
-	go Main()
+	smapp := newSpacemeshApp()
```
Comment (antonlerner) [inline app/cmd/app_test.go]: what is tested here? is this a working test?

File: app/cmd/hare.go (hunk 3 above)
Comment (gavraz) [reply app/cmd/hare.go]: proc.start is not blocking and shouldn't be blocking. The blocking part is in the caller of start which waits for the output.

File: app/cmd/base.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	"os"
+	"os/signal"
+	"reflect"
+)
+
+type baseApp struct {
+	Config *bc.Config
+}
+
+func newBaseApp() *baseApp {
```
Comment (antonlerner) [inline app/cmd/base.go]: rename

**Review by antonlerner — CHANGES_REQUESTED** [review]
refactor for better separation

**Pull request closed without merging**